flink-tests/src/test/java/org/apache/flink/test/checkpointing/RegionFailoverITCase.java

```diff
@@ -0,0 +1,392 @@
+							Thread.sleep(300);
+						}
+					}
+				}
+				if (jobFailedCnt.get() < NUM_OF_RESTARTS) {
+					// slow down if job has not failed for 'NUM_OF_RESTARTS' times.
+					Thread.sleep(1);
+				}
+			}
+		}
+
+		@Override
+		public void cancel() {
+			isRunning = false;
+		}
+
+		@Override
+		public void notifyCheckpointComplete(long checkpointId) {
+			if (getRuntimeContext().getIndexOfThisSubtask() == NUM_OF_REGIONS - 1) {
+				lastCompletedCheckpointId.set(checkpointId);
+				snapshotIndicesOfSubTask.put(checkpointId, lastRegionIndex);
+				numCompletedCheckpoints.incrementAndGet();
+			}
+		}
+
+		@Override
+		public void snapshotState(FunctionSnapshotContext context) throws Exception {
+			int indexOfThisSubtask = getRuntimeContext().getIndexOfThisSubtask();
+			if (indexOfThisSubtask != 0) {
+				listState.clear();
+				listState.add(index);
+				if (indexOfThisSubtask == NUM_OF_REGIONS - 1) {
+					lastRegionIndex = index;
+				}
+			}
+			unionListState.clear();
+			unionListState.add(indexOfThisSubtask);
+		}
+
+		@Override
+		public void initializeState(FunctionInitializationContext context) throws Exception {
+			int indexOfThisSubtask = getRuntimeContext().getIndexOfThisSubtask();
+			if (context.isRestored()) {
+				restoredState = true;
+
+				unionListState = context.getOperatorStateStore().getUnionListState(unionStateDescriptor);
+				Set<Integer> actualIndices = StreamSupport.stream(unionListState.get().spliterator(), false).collect(Collectors.toSet());
+				Assert.assertTrue(CollectionUtils.isEqualCollection(EXPECTED_INDICES, actualIndices));
+
+				if (indexOfThisSubtask == 0) {
+					listState = context.getOperatorStateStore().getListState(stateDescriptor);
+					Assert.assertTrue("list state should be empty for subtask-0",
+						((List<Integer>) listState.get()).isEmpty());
+				} else {
+					listState = context.getOperatorStateStore().getListState(stateDescriptor);
+					Assert.assertTrue("list state should not be empty for subtask-" + indexOfThisSubtask,
+						((List<Integer>) listState.get()).size() > 0);
+
+					if (indexOfThisSubtask == NUM_OF_REGIONS - 1) {
+						index = listState.get().iterator().next();
+						if (index != snapshotIndicesOfSubTask.get(lastCompletedCheckpointId.get())) {
+							throw new RuntimeException("Test failed due to unexpected recovered index: " + index +
+								", while last completed checkpoint record index: " + snapshotIndicesOfSubTask.get(lastCompletedCheckpointId.get()));
+						}
+					}
+				}
+			} else {
+				unionListState = context.getOperatorStateStore().getUnionListState(unionStateDescriptor);
+
+				if (indexOfThisSubtask != 0) {
+					listState = context.getOperatorStateStore().getListState(stateDescriptor);
+				}
+			}
+
+		}
+	}
+
+	private static class FailingMapperFunction extends RichMapFunction<Tuple2<Integer, Integer>, Tuple2<Integer, Integer>> {
+		private final int restartTimes;
+		private ValueState<Integer> valueState;
```

Comment:
Somehow I cannot find any code that validates if the key state is also properly restored, or is there something I missed?

Reply:
I'll send value as tuple of `<index/2, index>` from `StringGeneratingSourceFunction`, e.g. `1, 2` and `1, 3` would be emitted. `FailingMapperFunction` would then store keyed state `valueState` which tracks the maximum of value for the corresponding key, and emit `<key, max-value>` to `ValidatingSink`. In other words, we would have `1,3` finally in `ValidatingSink`'s `maps`. In `verifyAfterJobExecuted()` method, I'll then verify the `maps` in `ValidatingSink` works as expected to validate if key state has ever properly restored.